File: mobiletrm-backend/src/main/java/world/inetumrealdolmen/dto/TimeRegistrationDto.java
package world.inetumrealdolmen.dto;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import java.time.LocalDateTime;
import java.util.Set;
import world.inetumrealdolmen.domain.TimeRegistration;
import world.inetumrealdolmen.validation.NullOrNotBlank;
import world.inetumrealdolmen.validation.NullOrPositive;

/**
 * A collection of Data Transfer Objects (DTO)
 * for {@link TimeRegistration} entities to be sent to clients.
 * Use the static subclasses' constructors to create DTO's.
 */
public class TimeRegistrationDto {

  /**
   * Represents an index view of a time registration.
   */
  public static class Index {
    public long id;
    public String description;
    public LocalDateTime startTime;
    public LocalDateTime endTime;
    public TaskDto.Index assignedTask;
    public ProjectDto.Name assignedProject;
    public Set<TagDto.Index> tags;
  }

  /**
   * Represents a create request of a time registration.
   */
  public static class Create {
    @NullOrNotBlank
    public String description;
    @NotNull
    public LocalDateTime startTime;
    @NotNull
    public LocalDateTime endTime;
    @Positive
    public Long assignedProjectId;
    @NullOrPositive
    public Long assignedTaskId;
    public Set<TagDto.@Valid Index> tags;
  }

  /**
   * Represents an update request of a time registration.
   */
  public static class Update extends Create {
    @Positive
    public Long id;
  }

  /**
   * Represents the response upon creating a new time registration.
   *
   * @param id                the ID of the newly registered time registration.
   * @param description       the description of the newly registered time registration.
   * @param startTime         the startTime of the newly registered time registration.
   * @param endTime           the endTime of the newly registered time registration.
   * @param assignedProjectId the assigned project's id of the newly registered time registration.
   * @param assignedTaskId    the assigned task's id of the newly registered time registration.
   * @param tags              the assigned tags of the newly registered time registration.
   */
  public record CreateResponse(
      Long id,
      String description,
      LocalDateTime startTime,
      LocalDateTime endTime,
      Long assignedProjectId,
      Long assignedTaskId,
      Set<TagDto.Index> tags
  ) {
  }
}

File: mobiletrm-android/app/src/main/java/world/inetumrealdolmen/mobiletrm/data/dto/TimeRegistrationDTO.kt
package world.inetumrealdolmen.mobiletrm.data.dto

import kotlinx.datetime.LocalDateTime
import kotlinx.serialization.Serializable
import world.inetumrealdolmen.mobiletrm.data.model.TimeRegistration

/**
 * Detailed information of a time registration from the api.
 *
 * @param id The database ID of the time registration.
 * @param description The optional description of the time registration.
 * @param startTime The start time of the time registration.
 * @param endTime The end time of the time registration.
 * @param assignedTask The details of the optional task the time registration is for.
 * @param assignedProject The details of the project the time registration is for.
 * @param tags The list of optional tags for the time registration.
 */
@Serializable
data class TimeRegistrationDTO(
    val id: Long,
    val description: String?,
    val startTime: LocalDateTime,
    val endTime: LocalDateTime,
    val assignedTask: TaskDTO?,
    val assignedProject: ProjectDTO.IndexShort?,
    val tags: List<TagDTO>,
)

/**
 * Request body to send to the api to create a new time registration.
 *
 * @param description The optional description of the time registration.
 * @param startTime The start time of the time registration.
 * @param endTime The end time of the time registration.
 * @param assignedProjectId The database ID of the project the time registration is for.
 * @param assignedTaskId The database ID of the optional task the time registration is for.
 * @param tags The list of optional tags for the time registration.
 */
@Serializable
data class TimeRegistrationCreateDTO(
    val description: String = "",
    val startTime: LocalDateTime,
    val endTime: LocalDateTime,
    val assignedProjectId: Long,
    val assignedTaskId: Long? = null,
    val tags: List<TagDTO>? = listOf(),
)

/**
 * Request body to send to the api to update an existing time registration.
 *
 * @param id The database ID of the time registration.
 * @param description The optional description of the time registration.
 * @param startTime The start time of the time registration.
 * @param endTime The end time of the time registration.
 * @param assignedProjectId The database ID of the project the time registration is for.
 * @param assignedTaskId The database ID of the optional task the time registration is for.
 * @param tags The list of optional tags for the time registration.
 */
@Serializable
data class TimeRegistrationUpdateDTO(
    val id: Long? = null,
    val description: String = "",
    val startTime: LocalDateTime,
    val endTime: LocalDateTime,
    val assignedProjectId: Long,
    val assignedTaskId: Long? = null,
    val tags: List<TagDTO> = listOf(),
)

/**
 * Converts the response data to the [TimeRegistration] model.
 */
fun TimeRegistrationDTO.asDomainObject(): TimeRegistration =
    TimeRegistration(
        remoteId = id,
        description = description,
        startTime = startTime,
        endTime = endTime,
        assignedTask = assignedTask?.asDomainObject(),
        assignedProject = assignedProject,
        tags = tags.asDomainObjects(),
    )

/**
 * Converts the [TimeRegistration] model to a Create DTO to persist to the API.
 *
 */
fun TimeRegistration.asDto() =
    TimeRegistrationCreateDTO(
        description = description ?: "",
        startTime = startTime,
        endTime = endTime,
        assignedProjectId = assignedProject?.id ?: -1,
        assignedTaskId = assignedTask?.taskId,
        tags = tags.map { tag -> tag.asDto() },
    )
